fdbmonitor/fdbmonitor.cpp

```diff
@@ -223,34 +224,110 @@ std::string joinPath(std::string const& directory, std::string const& filename)
 	return d + CANONICAL_PATH_SEPARATOR + f;
 }
 
-std::string abspath(std::string const& filename) {
-	// Returns an absolute path canonicalized to use only CANONICAL_PATH_SEPARATOR
-	char result[PATH_MAX];
-	auto r = realpath( filename.c_str(), result );
-	if (!r) {
-		if (errno == ENOENT) {
-			int sep = filename.find_last_of( CANONICAL_PATH_SEPARATOR );
-			if (sep != std::string::npos) {
-				return joinPath( abspath( filename.substr(0, sep) ), filename.substr(sep) );
+std::string cleanPath(std::string const &path) {
+	std::vector<std::string> finalParts;
+	bool absolute = !path.empty() && path[0] == CANONICAL_PATH_SEPARATOR;
+
+	int i = 0;
+	while(i < path.size()) {
+		int sep = path.find((char)CANONICAL_PATH_SEPARATOR, i);
+		if(sep == path.npos) {
+			sep = path.size();
+		}
+		std::string part = path.substr(i, sep - i);
+		i = sep + 1;
+		if(part.size() == 0 || (part.size() == 1 && part[0] == '.'))
+			continue;
+		if(part == "..") {
+			if(!finalParts.empty() && finalParts.back() != "..") {
+				finalParts.pop_back();
+				continue;
 			}
-			else if (filename.find("~") == std::string::npos) {
-				return joinPath( abspath( "." ), filename );
+			if(absolute) {
+				continue;
 			}
 		}
+		finalParts.push_back(part);
+	}
 
-		log_err("realpath", errno, "Unable to get real path for %s", filename.c_str());
-		return "";
+	std::string result;
+	result.reserve(PATH_MAX);
+	if(absolute) {
+		result.append(1, CANONICAL_PATH_SEPARATOR);
 	}
-	return std::string(r);
+
+	for(int i = 0; i < finalParts.size(); ++i) {
+		if(i != 0) {
+			result.append(1, CANONICAL_PATH_SEPARATOR);
+		}
+		result.append(finalParts[i]);
+	}
+
+	return result.empty() ? "." : result;
 }
 
-std::string parentDirectory(std::string const& filename) {
-	size_t sep = filename.find_last_of( CANONICAL_PATH_SEPARATOR );
-	if (sep == std::string::npos) {
+// Removes the last component from a path string (if possible) and returns the result with one trailing separator.
+std::string popPath(const std::string &path) {
+	int i = path.size() - 1;
+	// Skip over any trailing separators
+	while(i >= 0 && path[i] == CANONICAL_PATH_SEPARATOR) {
+		--i;
+	}
+	// Skip over non separators
+	while(i >= 0 && path[i] != CANONICAL_PATH_SEPARATOR) {
+		--i;
+	}
+	// Skip over trailing separators again
+	bool foundSeparator = false;
+	while(i >= 0 && path[i] == CANONICAL_PATH_SEPARATOR) {
+		--i;
+		foundSeparator = true;
+	}
+
+	if(foundSeparator) {
+		++i;
+	}
+	else {
+		// If absolute then we popped off the only path component so return "/"
+		if(!path.empty() && path.front() == CANONICAL_PATH_SEPARATOR) {
+			return "/";
+		}
+	}
+	return path.substr(0, i + 1);
+}
+
+std::string abspath( std::string const& path, bool resolveLinks = true) {
+	if(path.empty()) {
 		return "";
 	}
 
-	return filename.substr(0, sep);
+	if(!resolveLinks) {
+		bool absolute = !path.empty() && path[0] == CANONICAL_PATH_SEPARATOR;
+		return cleanPath(absolute ? path : joinPath(abspath(".", true), path));
```

Comment:
It just occurred to me, should we also be dealing with `~` (in both locations)?

Reply:
Since `~` is not a thing in windows, I'd say we can simply return an error when see that?

Reply:
fdbmonitor is linux/mac only, and the version in flow can be run on any platform (although we disallow `resolveLinks==false` presently). I think it was also the case that we handled `~` previously.

Reply:
Yes this was dropped, I'll add it back.

Reply:
Actually, I don't think this was previously supported, just intelligently avoided.
https://github.com/apple/foundationdb/blob/master/flow/Platform.cpp#L1859
Using a ~ anywhere in the first path component (such as "a~b/c" or "~b/c") would cause it to NOT treat the path as relative and continue to the error case instead.  I'll restore this behavior.

Reply:
Fixed, but ~'s that are not the first character of a path component are not treated as special.  For reference, both MacOS and Linux both allow ~ to be an actual first character of a file or directory name, but we're making the reasonable assumption that that's not what the caller means if they have a ~ at the start of their path.